### ISSUE TITLE
no-useless-escape incorrectly reports errors for regular expressions with the unicode flag

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7789]
**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** 7.2.1
* **npm Version:** 3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

```yml
rules:
  no-useless-escape: error
parserOptions:
  ecmaVersion: 6
```

**What did you do? Please include the actual source code causing the issue.**

```js
/\]/u;
```

**What did you expect to happen?**

I expected no error to be reported. Although `]` does not need to be escaped outside of character classes when the unicode flag is disabled, it does need to be escaped when the unicode flag is enabled. It looks like a SyntaxError occurs when the backslash is removed in this example:

```js
/]/u; // SyntaxError
```

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint reported an error:

```
1:2  error  Unnecessary escape character: \]  no-useless-escape
```

Related: https://github.com/eslint/eslint/issues/7656

**Comment by not-an-aardvark** [comment]
See https://github.com/eslint/eslint/issues/7656#issuecomment-267878150

It looks like the best fix for this would be to just allow `\]` in all regular expressions, since `]` is only allowed to be unescaped in engines that implement Annex B.
